ntoskrnl/cc/copy.c

```diff
@@ -653,7 +653,9 @@ CcCopyWrite (
             CurrentOffset += VacbLength;
 
             /* Tell Mm */
-            Status = MmMakePagesDirty(NULL, Add2Ptr(Vacb->BaseAddress, VacbOffset), VacbLength);
+            Status = MmMakeSegmentDirty(FileObject->SectionObjectPointer,
+                                        Vacb->FileOffset.QuadPart + VacbOffset,
```

Comment:
> what is the formula that relate both quantities?

A VACB is mapped to it's corresponding byte offset when it is created (see `CcRosCreateVacb`).

The Mm functions (`MmArePagesResident` and `MmMakePagesDirty`) operate on the section segment with byte offset internally. They convert the supplied base address and length to their corresponding section segment with byte offset by finding it from the address space.

So the point of rewriting is: A VACB has the `FileOffset` field and the indirect link to it's section object pointer (via `Vacb->SharedCacheMap->FileObject->SectionObjectPointer`) already. Why not just get the section segment directly by calling `MiGrabDataSection` ? This already done in `MmMakeDataSectionResident` (also see `CcRosEnsureVacbResident`).

> is it possible to adjust the RtlCopyMemory call in a similar manner, or not?

Yes, it's possible, but that will over-complicate the code even more because `RtlCopyMemory` operates on a virtual address so we can use the `Vacb->BaseAddress` directly.